ZAP Result Scanner

ctro opened the pull request:
This ruby module summarizes (previous -> current) ZAP test results and outputs 2 things:
1) `summary.json`:  counts and summary for all projects
2) `summary.txt`: text summary for all projects

Much more info here: Ref 18F/staffing-and-resources#79

No deps, you can run the tests like this `ruby test_result_scanner.rb`
I didn't include a script that could just be called, but it'll look something like

``` rb
require_relative '../result_scanner'
ResultScanner.summarize_results(last_run_path, current_run_path, output_path)
```

Feedbacks?
- I ended up keeping an entire zap result-set around(micropurchase) in `test/fixtures`, so tests could run over a real dataset. Any issue w/ that?
- @DavidEBest I'm using the crude method of cache busing (dropping .js, .css on the ground).
- maybe it should be called ZapResultScanner? :)
- should I just include some exec-able script for use in concourse pipelines?

ctro committed:
ZAP Result Scanner
Ref 18F/staffing-and-resources#79

afeld commented on README.md:
````diff
@@ -11,10 +11,10 @@ The Concourse.ci site provides solid information for [Getting Started](http://co
+fly execute --config tasks/run-zap/task.yml --input scripts=. --output results=results
 ```
 
 This will use the current folder as the `scripts` input, and put the output designated by `results` to the folder of the same name.
````
Methinks we should just remove everything between "For example" and here, so that we don't have to keep the written-out command up-to-date in two places.

afeld commented on scripts/result_scanner/result_scanner.rb:
```diff
@@ -0,0 +1,118 @@
+    # Call Me.
+    def summarize_results(last_run_path, current_run_path, output_path)
+      @final_json_out = {}
+      @final_text_out = "Completed scan of #{json_files_from_dir(current_run_path).size} properties:"
```
Any reason these need to be an instance variables?

ctro replied:
Nope.  I think I `@`'d them just for syntax highlighting :stuck_out_tongue:
I'll fix.

afeld commented on scripts/result_scanner/test/test_result_scanner.rb:
```diff
@@ -0,0 +1,140 @@
+
+# For now, you should just be able to `ruby test_result_scanner.rb`
+
+class TestResultScanner < MiniTest::Test
```
Do you prefer MiniTest to RSpec?

ctro replied:
In general yes.  In my experience RSpec suites get more unwieldy over time than MiniTest.  MiniTest is part of ruby as of 2.0 (I think, maybe earlier), not a separate dependency. (But it can and _should_ be added to `.Gemfile` as a specific dep.  if/when that happens.)  MiniTest can do spec-style tests too.

afeld commented on README.md:
```diff
@@ -11,10 +11,10 @@ The Concourse.ci site provides solid information for [Getting Started](http://co
 After getting a local instance of Concourse up and running in Vagrant (by following the [Getting Started Guide](http://concourse.ci/getting-started.html)) and installing the Fly CLI, you will be set up for testing Concourse.
 
-Individual [tasks](http://concourse.ci/tasks.html) are run via the `fly execute` command, documented at the top of each `task.yml`. For example, you can run `zap-task` via the following:
+Individual [tasks](http://concourse.ci/tasks.html) are run via the `fly execute` command, documented at the top of each `task.yml`. For example, you can run Zap via the following:
```
Minor: ZAP is written in all caps.

afeld commented on scripts/result_scanner/result_scanner.rb:
```diff
@@ -0,0 +1,118 @@
+
+  end #/ class << self.
+
+end #/ module
```
This module is getting kinda long already...maybe we make a class for `Result`, and another for `ResultComparator` or something? Or they could be modules if you prefer the more functional style.

afeld commented on scripts/result_scanner/result_scanner.rb:
```diff
@@ -0,0 +1,118 @@
+
+        @final_json_out.merge!({ proj => (count_risk_levels(current_results).merge(status: project_status)) })
+        @final_text_out << "\n#{proj}: #{paren_status_count(current_results)} #{project_status}"
+      end
```
Can we simplify this loop?
- Move things into smaller methods
- Use more variables rather than nesting function calls

ctro replied:
I was daydreaming of Elixir's pipe (`|>`) operator :)

afeld commented on scripts/result_scanner/result_scanner.rb:
```diff
@@ -0,0 +1,118 @@
+    def compute_risk_level_deltas(old_results, new_results)
+      deltas = {}
+      deltas[:decreased_counts] = count_risk_levels(old_results - new_results)
+      deltas[:increased_counts] = count_risk_levels(new_results - old_results)
```
Minor: I'd create the Hash and assign these values in one go.

``` ruby
return {
  decreased_counts: ...,
  increased_counts: ...
}
```

afeld commented on scripts/result_scanner/result_scanner.rb:
```diff
@@ -0,0 +1,118 @@
+
+    # We are dropping all Results with `url` of .css or .js as a crude way
+    #  of dealing with cache-busting urls.
+    def is_js_or_css?(url)
```
Minor: I'd leave the `is_` off the method name.

afeld commented on scripts/result_scanner/test/test_result_scanner.rb:
```diff
@@ -0,0 +1,140 @@
+  end
+
+  def test_a_full_result_set
+    ResultScanner.summarize_results("#{__dir__}/fixtures/last_full_mp", "#{__dir__}/fixtures/current_full_mp", __dir__)
```
I'd suggest using the block form of [`Dir.mktmpdir`](http://ruby-doc.org/stdlib-2.0.0/libdoc/tmpdir/rdoc/Dir.html#method-c-mktmpdir) for the `output_path` so that it does the cleanup automatically.

ctro replied:
Good idea!

afeld commented on scripts/result_scanner/test/test_result_scanner.rb:
```diff
@@ -0,0 +1,140 @@
+    @last_json = ResultScanner.read_json("#{@last_run_path}/fake-site-1.json")
+    @current_json = ResultScanner.read_json("#{@current_run_path}/fake-site-1.json")
+    @last_results = ResultScanner.create_results("fake-site-1", @last_json)
+    @current_results = ResultScanner.create_results("fake-site-1", @current_json)
```
Not a blocker, but we might want to use a Factory for most of these so that our test data isn't as stagnant, especially for the more specific unit tests.

ctro replied:
Yes.  I've used FactoryGirl in the past.  CT needs a "real" test suite that can load deps for ruby (and js and python?). I'm not sure what that looks like (?). And once that happens these tests will require some more work.  They mostly exist as dev artifacts atm.

afeld commented:
Nice work! I imagine this will move beneath a particular Task at some point, but fine if we keep it in `scripts` for now.

ctro committed:
Addressed Feedback

Thanks @afeld for all the great feedback, it made things better.

- prune README
- fix unneeded class vars
- split out ResultScanner
- simplify the main `write_summary` loop
- better temp files
- better names
- some simlification & other odds and ends

ctro commented on scripts/result_scanner/result_scanner.rb:
```diff
@@ -0,0 +1,118 @@
+      return Dir["#{path}/*.json"]
+    end
+
+    def write_file(file, contents)
```
haha I just noticed this unused method, oops!

ctro commented:
Oof, that last commit didn't rename a file :(  Next time I'll be more careful to add moving commits before doing other stuff.
The full diff still looks OK tho.

afeld commented on scripts/zap_result_scanner/lib/zap_result.rb:
```diff
@@ -0,0 +1,50 @@
+
+    def read_json(path)
+      return File.exists?(path) ? JSON.parse(File.read(path)) : []
+    end
```
This seems like it should give an error if the path doesn't exist...right?

ctro replied:
it's meant to return an actual JSON obj, not a string.   According to the Ask missing files could be common, in the script it's useful to think of a missing results file as producing an empty Result list ([]).  Maybe the method name could be more clear?

afeld replied:
Yeah, I'd throw a comment in there at least.

afeld commented on scripts/zap_result_scanner/lib/zap_result_comparator.rb:
```diff
@@ -0,0 +1,66 @@
+
+        final_json_out.merge!({ proj => project_summary })
+        final_text_out << "\n#{proj}: #{paren_status_count(curr_results)} #{project_status}"
+      end
```
Can we break this loop down into smaller functions/classes?

afeld commented on scripts/zap_result_scanner/test/test_zap_result_scanner.rb:
```diff
@@ -0,0 +1,134 @@
+  end
+
+  def test_a_full_result_set
+    Dir.mktmpdir do |dir|
```
Maybe change `dir` to `results_dir`, just to differentiate it from `__dir__`? Ditto below.

afeld commented:
Just a couple small things. Thanks for all the adjustments!

ctro committed:
Address Feedback #2

- Now with an even slimmer `write_summary` loop
- A helpful comment and rename.

DavidEBest commented:
:+1:
